regexp/regexp.go

```diff
@@ -0,0 +1,110 @@
+package regexp
+
+import (
+	"regexp"
+	"runtime"
+	"sync"
+	"weak"
+)
+
+// "Interns" compilation of Regular Expressions.  If two regexs with the same pattern are compiled, the result
+// is the same *regexp.Regexp.  This avoids the compilation cost but more importantly the memory usage.
+//
+// Regular expressions produced from this package are backed by a form of weak-valued map, upon a regexp becoming
+// unreachable, they will be eventually removed from the map and memory reclaimed.
+
+var (
+	weakMap      = make(map[string]weak.Pointer[regexp.Regexp])
+	posixWeakMap = make(map[string]weak.Pointer[regexp.Regexp])
+	reverseMap   = make(map[weak.Pointer[regexp.Regexp]]string)
+	l            sync.RWMutex
+)
+
+// CompileInterned compiles and interns a regular expression and returns a
+// pointer to it or an error.
+func CompileInterned(pattern string) (*regexp.Regexp, error) {
+	return compile(pattern, regexp.Compile, weakMap)
+}
+
+// CompilePOSIXInterned compiles and interns a regular expression using POSIX
+// syntax and returns a pointer to it or an error.
+func CompilePOSIXInterned(pattern string) (*regexp.Regexp, error) {
+	return compile(pattern, regexp.CompilePOSIX, posixWeakMap)
+}
+
+// MustCompileInterned compiles and interns a regular expression and returns a pointer to it or panics.
+func MustCompileInterned(pattern string) *regexp.Regexp {
+	return mustCompile(pattern, regexp.MustCompile, weakMap)
+}
+
+// MustCompilePOSIXInterned compiles and interns a regular expression using
+// POSIX syntax and returns a pointer to it or panics.
+func MustCompilePOSIXInterned(pattern string) *regexp.Regexp {
+	return mustCompile(pattern, regexp.MustCompilePOSIX, posixWeakMap)
+}
+
+// compile handles compiling and interning regular expressions. If the regexp is
+// already interned, a pointer to it is returned from the map. If the regexp is
+// not interned or is nil (since it's a weak pointer), it is compiled and stored
+// in the maps. The regexp is stored in the maps as a weak pointer, so that it
+// can be garbage collected.
+func compile(pattern string, compileFunc func(string) (*regexp.Regexp, error), weakMap map[string]weak.Pointer[regexp.Regexp]) (*regexp.Regexp, error) {
+	l.Lock()
```

Comment:
This single-lock approach can create a lot of unnecessary contention if a bunch of disjoint compilations are triggered simultaneously right? I wonder if we should consider something like [singleflight](https://pkg.go.dev/github.com/golang/groupcache/singleflight) to allow multiple compilations of different patterns to happen at the same time.

Reply:
yeah a whole other dependency might be a bit much, but I think we could address it in other ways. For instance, we could compile the regext without holding the mutex, then later lock it and check if another goroutine has compiled the same pattern in which case we return the existing compilation, otherwise we add it to the map. That would allow multiple disjoint compilations to not contend on the lock since the most time consuming step is removed from the critical section. The drawback is that in case of parallel compilations of the same patterns we'd be wasting CPU, but I think that is fine since the main goal here is to not waste RAM which would be accomplished by discarding the duplicate compilation.

Reply:
Technically we only really need the lock around the map checking/updating, but my hunch is that it doesn't matter either way. I doubt leaving it the way it is will introduce enough lock contention that it will result in a performance hit we can notice, and similarly I doubt that moving the lock to the critical section around map checking/writing will result in a performance increase that we can notice.

Reply:
I am going to resolve this thread for now. If we ever observe any issues related to lock contention, we can come back and address it.